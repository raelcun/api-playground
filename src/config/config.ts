import deepmerge from 'deepmerge'

import { PartialConfig } from './types'

export type ConfigMap<T> = { default: T; [env: string]: PartialConfig<T> | undefined }

const merge = <T>(config: T, partials: PartialConfig<T>[]): T => {
  const overwriteMerge = (destination: any[], source: any[]) => source

  let resConfig: T = config
  partials.forEach(partial => {
    resConfig = deepmerge<T, PartialConfig<T>>(resConfig, partial, {
      arrayMerge: overwriteMerge,
    })
  })

  return resConfig
}

export const mergeConfig = <T>(configMap: ConfigMap<T>, env: string, context: string) => {
  const partialKeys = [
    `default-${context}`,
    env,
    `local-${env}`,
    `${env}-${context}`,
    `local-${env}-${context}`,
  ]

  const partials = partialKeys
    .map(key => configMap[key])
    .filter((e: PartialConfig<T> | undefined): e is PartialConfig<T> => e !== undefined)

  return merge(configMap.default, partials)
}
