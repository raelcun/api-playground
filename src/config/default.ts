import { FullConfig } from './types'

const config: FullConfig = {
  server: {
    port: 4443,
  },
}

module.exports = {
  ...config,
  default: config,
}
