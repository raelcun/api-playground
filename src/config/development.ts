import { PartialConfig } from './types'

const config: PartialConfig = {
  server: {
    enableSSL: false
  }
}

module.exports = {
  ...config,
  default: config,
}
