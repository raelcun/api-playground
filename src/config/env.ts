import { either as E } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'

import { getEmergencyLogger } from '@modules/emergencyLogger'
import { Err } from '@modules/error/types'

import { EnvVars, envVarsV } from './types'

export const getEnvVars = (): E.Either<Err, EnvVars> =>
  pipe(
    envVarsV.decode({
      NODE_ENV: process.env.NODE_ENV,
      JWT_SECRET: process.env.JWT_SECRET || '',
      AWS_ACCESS_KEY_ID: process.env.AWS_ACCESS_KEY_ID || '',
      AWS_SECRET_ACCESS_KEY: process.env.AWS_SECRET_ACCESS_KEY || '',
    }),
    E.mapLeft(errors => {
      getEmergencyLogger().fatal('environment variable validation failed', errors)
      return { code: 'ENV_VALIDATION_FAILED' }
    }),
  )
