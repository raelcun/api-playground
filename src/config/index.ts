import { either as E } from 'fp-ts'

import { getEmergencyLogger } from '@modules/emergencyLogger'

import { ConfigMap, mergeConfig } from './config'
import getDefaultConfig from './configuration/default'
import getDevelopmentConfig from './configuration/development'
import getDevelopmentTestingConfig from './configuration/development-testing'
import getProductionConfig from './configuration/production'
import { FullConfig } from './types'

let config: FullConfig
const loadConfig = () => {
  const configMap: ConfigMap<FullConfig> = {
    default: getDefaultConfig,
    development: getDevelopmentConfig,
    'development-testing': getDevelopmentTestingConfig,
    production: getProductionConfig,
  }

  const mergeResult = mergeConfig(
    configMap,
    process.env.NODE_ENV || '',
    process.env.NODE_APP_CONTEXT || '',
  )

  if (E.isLeft(mergeResult)) {
    getEmergencyLogger().fatal('CONFIG_LOAD_FAILED', mergeResult.left)
    throw mergeResult.left
  } else {
    config = mergeResult.right
  }
}

export const getConfig = (): FullConfig => {
  if (config === undefined) {
    loadConfig()
  }
  return config
}
