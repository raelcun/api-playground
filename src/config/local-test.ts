import { PartialConfig } from './types'

const config: PartialConfig = {
  logging: {
    level: 'error',
  },
}

module.exports = {
  ...config,
  default: config,
}
