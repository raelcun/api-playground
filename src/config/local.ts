import { PartialConfig } from './types'

const config: PartialConfig = {
  logging: {
    level: 'trace',
  },
}

module.exports = {
  ...config,
  default: config,
}
