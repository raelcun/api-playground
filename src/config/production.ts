import { PartialConfig } from './types'

const config: PartialConfig = {
  server: {
    port: 4443,
  },
  logging: {
    level: 'error',
  },
}

module.exports = {
  ...config,
  default: config,
}
