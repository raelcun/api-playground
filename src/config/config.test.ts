import { filterObjectKeys } from '@utils'

import { ConfigMap, mergeConfig } from './config'

type MockConfig = {
  a: {
    b: number
    c: string
    arr: {
      d: {
        e: string
        f?: string
      }
    }[]
  }
  g: number
}

const defaultConfig: MockConfig = {
  a: {
    b: 5,
    c: 'c',
    arr: [
      {
        d: {
          e: 'e',
          f: 'f',
        },
      },
      {
        d: {
          e: 'e',
        },
      },
    ],
  },
  g: 8,
}

describe('config', () => {
  test('overrides nested object keys', () => {
    const configMap: ConfigMap<MockConfig> = {
      default: defaultConfig,
      'default-test': {
        a: {
          c: 'newC',
        },
      },
    }

    expect(mergeConfig(configMap, 'env', 'test')).toMatchSnapshot()
  })

  test('overrides whole array', () => {
    const configMap: ConfigMap<MockConfig> = {
      default: defaultConfig,
      'default-test': {
        a: {
          arr: [
            {
              d: {
                e: 'foo',
              },
            },
          ],
        },
      },
    }

    expect(mergeConfig(configMap, 'env', 'test')).toMatchSnapshot()
  })

  test('missing partial config does not fail', () => {
    const configMap: ConfigMap<{ a: number }> = {
      default: { a: 1 },
      stage: { a: 2 },
    }

    expect(mergeConfig(configMap, '', '')).toMatchSnapshot()
  })

  describe('config hierarchy', () => {
    const env = 'stage'
    const context = 'testing'
    const keyMap = {
      defaultContext: `default-${context}`,
      env,
      localEnv: `local-${env}`,
      envContext: `${env}-${context}`,
      localEnvContext: `local-${env}-${context}`,
    }
    const fullConfigMap: ConfigMap<{ foo: number }> = {
      default: { foo: 1 },
      [keyMap.defaultContext]: { foo: 2 },
      [keyMap.env]: { foo: 3 },
      [keyMap.localEnv]: { foo: 4 },
      [keyMap.envContext]: { foo: 5 },
      [keyMap.localEnvContext]: { foo: 6 },
    }

    test.each([
      ['defaultContext', [keyMap.defaultContext]],
      ['env', [keyMap.defaultContext, keyMap.env]],
      ['localEnv', [keyMap.defaultContext, keyMap.env, keyMap.localEnv]],
      ['envContext', [keyMap.defaultContext, keyMap.env, keyMap.localEnv, keyMap.envContext]],
      [
        'localEnvContext',
        [
          keyMap.defaultContext,
          keyMap.env,
          keyMap.localEnv,
          keyMap.envContext,
          keyMap.localEnvContext,
        ],
      ],
    ])('%s should take highest priority', (_, keys) => {
      expect(
        mergeConfig(
          {
            default: fullConfigMap.default,
            ...filterObjectKeys(fullConfigMap, keys),
          },
          env,
          context,
        ),
      ).toMatchSnapshot()
    })
  })
})
