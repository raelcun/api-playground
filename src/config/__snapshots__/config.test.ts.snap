// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config config hierarchy defaultContext should take highest priority 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "foo": 2,
  },
}
`;

exports[`config config hierarchy env should take highest priority 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "foo": 3,
  },
}
`;

exports[`config config hierarchy envContext should take highest priority 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "foo": 5,
  },
}
`;

exports[`config config hierarchy localEnv should take highest priority 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "foo": 4,
  },
}
`;

exports[`config config hierarchy localEnvContext should take highest priority 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "foo": 6,
  },
}
`;

exports[`config failed default config fails merge 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "VALIDATION_FAILED",
  },
}
`;

exports[`config failed partial config fails merge 1`] = `
Object {
  "_tag": "Left",
  "left": Object {
    "code": "VALIDATION_FAILED",
  },
}
`;

exports[`config missing partial config does not fail 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "a": 1,
  },
}
`;

exports[`config overrides nested object keys 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "a": Object {
      "arr": Array [
        Object {
          "d": Object {
            "e": "e",
            "f": "f",
          },
        },
        Object {
          "d": Object {
            "e": "e",
          },
        },
      ],
      "b": 5,
      "c": "newC",
    },
    "g": 8,
  },
}
`;

exports[`config overrides whole array 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "a": Object {
      "arr": Array [
        Object {
          "d": Object {
            "e": "foo",
          },
        },
      ],
      "b": 5,
      "c": "c",
    },
    "g": 8,
  },
}
`;
